java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/debugging/attach/NbAttachRequestHandler.java

```diff
@@ -0,0 +1,217 @@
+        VirtualMachineManager vmm = Bootstrap.virtualMachineManager ();
+        List<AttachingConnector> attachingConnectors = vmm.attachingConnectors();
+        for (AttachingConnector connector : attachingConnectors) {
+            if (connector.name().equals(connectorName)) {
+                Map<String, Argument> args = connector.defaultArguments();
+                for (String argName : arguments.keySet()) {
+                    if (CONFIG_ATTRIBUTES.contains(argName) || argName.startsWith("__")) {
+                        continue;
+                    }
+                    String argNameTranslated = translatedArguments.getOrDefault(argName, argName);
+                    Argument arg = args.get(argNameTranslated);
+                    if (arg == null) {
+                        ErrorUtilities.completeExceptionally(resultFuture,
+                            "Argument " + argNameTranslated + " of " + connectorName + " was not found.",
+                            ResponseErrorCode.serverErrorStart);
+                        return ;
+                    }
+                    String value = arguments.get(argName).toString();
+                    if (!arg.isValid(value)) {
+                        ErrorUtilities.completeExceptionally(resultFuture,
+                            "Invalid value of " + argName + ": " + value,
+                            ResponseErrorCode.serverErrorStart);
+                        return ;
+                    }
+                    arg.setValue(value);
+                }
+                AttachingDICookie attachingCookie = AttachingDICookie.create(connector, args);
+                resultFuture.complete(null);
+                startAttaching(attachingCookie, context);
+                return ;
+            }
+        }
+        ErrorUtilities.completeExceptionally(resultFuture,
+                "Connector " + connectorName + " was not found.",
+                ResponseErrorCode.serverErrorStart);
+    }
+
+    private void startAttaching(AttachingDICookie attachingCookie, DebugAdapterContext context) {
+        DebuggerEngine[] es = DebuggerManager.getDebuggerManager ().startDebugging(
+            DebuggerInfo.create(AttachingDICookie.ID, new Object [] { attachingCookie })
+        );
+        if (es.length > 0) {
+            JPDADebugger debugger = es[0].lookupFirst(null, JPDADebugger.class);
+            if (debugger != null) {
+                Session session = es[0].lookupFirst(null, Session.class);
+                NbDebugSession debugSession = new NbDebugSession(session);
+                context.setDebugSession(debugSession);
+                AtomicBoolean finished = new AtomicBoolean(false);
+                debugger.addPropertyChangeListener(JPDADebugger.PROP_STATE, new PropertyChangeListener() {
+                    private final AtomicBoolean initialized = new AtomicBoolean(false);
```

Comment:
field seems to be unused.

Reply:
Thanks! This is a relict from the initial version.